### DISCUSSION TITLE
Centralized downloaded bytes accumulator

### DISCUSSION THREAD

**Discussion by gcarreno** [mr-highball/ezthreads#13] [Q&A] [answered]
Hey @mr-highball,

To have a thread pool downloading many files in parallel, but one single place for the total bytes, should I use an external critical section, OR, does your pool manager provide such service?

Cheers,
Gus

**Comment by mr-highball** [comment, marked as answer]
So let me hopefully answer your question by first asking my own questions.  Which way do you plan on using the pool?
1. instantiate a pool at the start of your program and download lots of files for the lifetime of the program
2. instantiate a pool at the time you need to download lots of files and discard the pool once they're finished download

Currently there is a route to access the parent pool from within the worker thread method, which supports `.AddArg(someName, someArgument)`, however it is undocumented but could be added pretty quickly (this would support scenario 2 since you could await the entire pool, then read the contents like this once all files have finished.
```pascal
Pool.Queue(job(s))
Await(Pool)
FIle1 := Pool['file1'];
FIle2 := Pool['file2'];
```

this approach would also need to make use of variants, so for bytes, it would probably be best to just use a string for the contents.

Now, back to my first approach, I could expose a callback to add to the queue for when a single worker finishes. This could allow you to call `AWorker.AddArg[file...` and then when the OnWorkerStop it could respect the "synchronize" settings on the pool so you could rely on the internal critical section. This approach also requires the use of variants, so a third approach is to use your own critical section and storage data structure and just handle the writing to this inside of the worker method. 

thoughts?

**Reply by gcarreno** [reply]
Ok, I think I need to make my context a bit more clear.

1. You click on a button to refresh a set of files that are the core data for your app, specifically [corona](https://github.com/wp-xyz/corona).
2. You do a HEAD of the files to get the total Bytes and store it somewhere above pool manager or in pool manager
3. For each file you queue a thread
4. Using `TfpHTTPClient` and it's call back `OnDataReceive` you get the position and total bytes
5. You synchronize somewhere to make the progress bar show the current status.
6. If you got to [this issue](https://github.com/wp-xyz/corona/issues/16), on the first comment I dropped an example with two options: Sequential non threaded and Sequential threaded. What I want, but didn't provide, is Async threaded.

Werner(@wp-xyz), decided to implement Sequential threaded [here: cdownloadmanager.pas](https://github.com/wp-xyz/corona/blob/master/source/cdownloadmanager.pas)

Now, I wasn't asking you to solve my issue. I was asking if you have a tidy way of doing this already.
If you don't, please, oh please don't worry, I'll manage ok with a critical section around updating the total bytes counter.

I was just curious if you encountered a similar usage of a pool that would need something central to all the thread on a pool.

Hope that makes it clearer. Again sorry for being so vague, wasn't trying anything funny, just aiming for generic, ehehe

Cheers,
Gus

**Reply by mr-highball** [reply]
A use case gives me a good bit more to go off of, and I'll put together a simple demo how I'd approach it with the pool. won't take me too long and it would be good to show off how to use the library. will post back a little later today once it's done

**Reply by gcarreno** [reply]
Hey @mr-highball,

That would be doubly awesome of you!! Thanks so much!!!

Cheers,
Gus

**Reply by mr-highball** [reply]
didn't finish up had some things to do today, but I got a decent start and will finish up tomorrow sometime, but here's the demo location if you want to start looking over how I'm approaching it (again... still work in progress)
https://github.com/mr-highball/ezthreads/blob/master/demo/pool-multi-file-download/main.pas

really the only thing that should be left is just wiring up the progress bar's DataReceived

**Reply by mr-highball** [reply]
@gcarreno I ended up finishing this and uploaded my demo here
https://github.com/mr-highball/ezthreads/tree/master/demo/pool-multi-file-download

gif showing the usage. the demo actually let me catch a nasty deadlock bug with the pool so hopefully that should be all sorted out now 🤞. let me know if you have any ?'s

you can pretty much use the multi file download form "as is" if desired (don't even have to show it to user) since there are properties for the files to download, the file data and clear/download methods. the only thing missing would be just a better way to notify when everything is finished, but there is a stubbed out method for "finished" that you change however you'd like (some callback, write some property, whatever...)
https://github.com/mr-highball/ezthreads/blob/819c650992c4e5022b732058a5a6fdeb7f941eba/demo/pool-multi-file-download/main.pas#L243

**Reply by gcarreno** [reply]
Hey @mr-highball,

The demo is awesome, thank you so very much!!!!

I've had a diagonal read of the code but I want to run it here to have a better sense. It's getting late here and I'll do it tomorrow.

> The demo actually let me catch a nasty deadlock bug with the pool so hopefully that should be all sorted out now :crossed_fingers:.

Awesome, that's what we want!!!

> but there is a stubbed out method for "finished" that you change however you'd like (some callback, write some property, whatever...)

Yeah, inside `FinishedDownloads()` whe can can synchronize something...

Cheers,
Gus

PS: My eye-lids are getting very heavy and my brain is not quite functioning. Sorry if I'm a bit vague :smile:

**Reply by mr-highball** [reply]
No problem, it's also getting late here. Actually I think the demo is somewhat incomplete with the progress bar implementation. I need to think a little bit on a good way to get a snapshot of the current progress. Think I have an idea but will sleep on it. Also the finish download method is already guaranteed to be in the context of the main thread so anything can be done there without synchronize.

**Reply by gcarreno** [reply]
> I need to think a little bit on a good way to get a snapshot of the current progress.

Okydokes, then I'll be eager to have a look at it when you finish!!

When I was thinking about it I had 2 ideas:
1. Have a `totalBytes`(HEAD of all files) and a `totalProgress` that gets incremented after each `DataReceived()` callback, and then
```pascal
begin
  {...}
  // Since Position is Integer and not Int64, if you download a 4GB file, not issues
  ProgressBar.Position:= (totalProgress * 100) / totalBytes;
  {...}
end;
```
2. Instead of decrementing `totalProgess`, decrement `totalBytes`?
  - I know, it came to my mind, but now that I write it down, doesn't makes a lot of sense...


> Also the finish download method is already guaranteed to be in the context of the main thread so anything can be done there without synchronize.

Makes sense!

**Reply by mr-highball** [reply]
@gcarreno finished up the demo, but let me know if you have any ?'s since there's a decent amount of stuff going on there. tried to comment pretty well so hopefully it's easy enough to follow

**Reply by gcarreno** [reply]
Hey @mr-highball ,

Thanks, so very much, for the effort!!

Will have a deeper look.

Cheers,
Gus
